HPCC-30037 Check legacy DZ physical permission in ESP services

wangkx opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx commented:
@jakesmith please review the second commit which is based on the commit in https://github.com/hpcc-systems/HPCC-Platform/pull/17528.

jakesmith commented:
@wangkx - before I review, now that https://track.hpccsystems.com/browse/HPCC-29246 is merged, can you rebase and change from draft status?

jakesmith reviewed (commented):
@wangkx - please rebase

wangkx commented:
Will do rebase

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -2013,15 +2013,12 @@ void CFileSprayEx::readAndCheckSpraySourceReq(IEspContext& context, MemoryBuffer
-            }
+            //Based on the tests, the dfuserver only supports the wildcard inside the file name, like '/path/f*'.
+            //The dfuserver throws an error if the wildcard is inside the path, like /p*ath/file.
+            SecAccessFlags permission = getDZFileScopePermissions(context, sourcePlaneReq, path, sourceIPReq, SecAccess_Read);
```
Removing the check on if (dropZone) line 2016 (and elsewhere throughout these changes), doesn't look right.

It is conflating 2 different issues:
1) The existence of a valid dropzone definition in the environment (or no match if restrictions are off)
2) LDAP checking of that dropzone based on PLANE::<DROPZONE>::... or failing over to legacy in bare-metal if failover is configured.

If the dropzone in 1) doesn't exist (and was allowed to continue because restrictions off), then it should not continue to scope check.
If the dropzone does exist, it should scope check and that scope checking should failover to legacy physical form if failOverToLegacyPhysicalPerms is on.

NB: it is also wrong in dfurun.cpp, it too should not go on to check if an environment dropzone couldn't be found (and was allowed to proceed because restrictions were off).

jakesmith replied:
only if dropzone exists. And as noted above, dfurun.cpp should do same.

wangkx replied:
How about the Foreign File in dfurun.cpp? Should I check legacy physical form there if the failOverToLegacyPhysicalPerms is on?

jakesmith replied:
yes, that is not dependent on whether there's a dropzone or not.
The comments above are only related to what to do when there's no dropzone found (which implies restrictions off too).

jakesmith reviewed (changes requested):
@wangkx - see comment.

wangkx committed:
HPCC-30037 Check legacy DZ physical permission in ESP services

1. In checkPlaneFilePermissions() (dfurun.cpp), check legacy
physical permission only if the plane name is set.
2. Add getLegacyDZPhysicalPerms() in ESP SMCLib.
3. Call getLegacyDZPhysicalPerms() in validateDropZoneAccess()
which is used by ESP WsFileIO service.
4. In ESP FileSpray service, check legacy physical permission
if plane exists.

Revise based on review
1. Add queryDistributedFileDirectory().getFScopePermissions() and
call it in the getLegacyDZPhysicalPerms().
2. Move validateDropZoneReq(), validateDZFileScopePermission()
and getDZPathScopePermissionsEx()) into TpCommon.cpp.
Rename the getDZPathScopePermissionsEx() to
getDZPathScopePermsAndLegacyPhysicalPerms().
Rename the validateDZFileScopePermission() to
validateDZFileScopePermissions(). Also revise
them.
3. The following 3 files are not called from outside. Change them
from extern to static: getDZPathScopePermissions(),
getDZFileScopePermissions() and getLegacyDZPhysicalPerms().
4. Add comments.
5. Fix a bug in the message for the throwOrLogDropZoneLookUpError().
6. Rename the validateDropZoneHostAndPath() to getDropZoneAndValidateHostAndPath()
and revise it. It is called when uploading/downloading dropzone file and in
WsFileSprayEx::onFileList(). There are 2 issues in the existing code: a. some
code are duplicated with the code in FileSprayEx::onFileList(). b. the
isDropZoneRestrictionEnabled is not used.
7. Rename the validateDZFileScopePermissions() to
   validateDZFileScopePermsAndLegacyPhysicalPerms()
8. Consoliate the code in FileSprayEx::onFileList() by using the
getDropZoneAndValidateHostAndPath().
9. Add validateDZPathScopePermsAndLegacyPhysicalPerms() which contains the common
code for the validateDropZoneReq() and WsFileSprayEx.
10. Revise the code in validateDropZoneReq().
11. Move the addPathSepChar() line, fix an intent bug, and fix a message bug.
12. Add more comments.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith please review (I do not keep the previous commit since it is fully changed).

jakesmith commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -230,6 +230,37 @@ extern TPWRAPPER_API SecAccessFlags getDZFileScopePermissions(IEspContext& conte
+
+    //The setExternal() requests the path with a file name. The file name is not used by the getDLFNPermissions().
+    //If the dropZoneFile is a path only, the dropZoneScope + a faked file name is passed to the setExternal().
+    StringBuffer pathWithFileName(dropZoneFile);
```
it would help if there were 2 functions one for Path and one for File (getLegacyDZPhysicalFilePerms, getLegacyDZPhysicalPathPerms)
And, rather than this workaround for the lack of support from IDistributedFileDirectory for validating a path that isn't a file, it would be better to add support.
What's missing I think is a simple:
```
IDistributedFileDirectory::getScopePermissions(const char *scopes, ....) 
```

method.

If that is added, then the common code for that getLegacyDZPhysicalFilePerms and getLegacyDZPhysicalPathPerms calls , can use a CDfsLogicalFileName and conditionally do:
```
if (isPath)
    scopesToCheck = dlfn.get();
else
{
    dlfn.getScopes(scopes);
    scopesToCheck = scopes;
}
return queryDistributedFileDirectory().getScopePermissions(scopesToCheck, userDesc);

```

jakesmith reviewed (changes requested):
@wangkx - please see comments

wangkx commented on esp/services/ws_fs/ws_fsBinding.cpp:
```diff
@@ -412,12 +425,7 @@ int CFileSpraySoapBindingEx::downloadFile(IEspContext &context, CHttpRequest* re
-        if (permission < SecAccess_Read)
-            throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone Scope %s %s %s not allowed for user %s (permission:%s). Read Access Required.",
-                dropZoneName.str(), netAddressStr.str(), pathStr.str(), context.queryUserId(), getSecAccessFlagName(permission));
+        validateDropZoneReq(context, dropZoneName, netAddressStr, pathStr, SecAccess_Read);
```
The validateDropZoneReq() calls the getDZPathScopePermissions() which handles the isDropZoneRestrictionEnabled if a dropzone cannot be found based on host and path (a bug to be fixed inside the error message). All other DZ permission checks in ESP should go through the getDZPathScopePermissions().
Inside the getDZPathScopePermissions(), if a dropzone name is given, but, a dropzone cannot be found based on the name, an exception is thrown. Should the throwOrLogDropZoneLookUpError be called instead?

wangkx commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -1920,6 +1920,16 @@ static bool parseUNCPath(const char* sprayPath, StringBuffer& localPath, StringB
     return true;
 }
 
+static void validateDZFileScopePermission(IEspContext& context, const char* dropZoneName, const char* path, const char* host, SecAccessFlags permissionReq)
```
This function calls getDZFileScopePermissions() and the new function in ws_fsBinding.cpp calls getDZPathScopePermissions(). I am going to move (possible rename and merge) them (and the getDZPathScopePermissionsEx()) into TpCommon.cpp.

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -3477,8 +3471,8 @@ bool CFileSprayEx::onDropZoneFiles(IEspContext &context, IEspDropZoneFilesReques
-        if (!isEmptyString(dzName) && getDZPathScopePermissions(context, dzName, directoryStr, nullptr) < SecAccess_Read)
-            return false;
+        if (!isEmptyString(dzName) && (getDZPathScopePermsAndLegacyPhysicalPerms(context, dzName, directoryStr, netAddress, SecAccess_Read) < SecAccess_Read))
+                return false;
```
odd indentation

wangkx replied:
Will fit it

jakesmith reviewed (commented):
@wangkx - looks better, a couple of questions surrounding useDropZoneRestriction in the case of upload/download. 
Could still do with more clarification (comments) on the different but very similar functions in TpCommon.cpp.

And, I think there's more scope to rationalize them further, but that can wait for another day.

wangkx commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -230,6 +231,77 @@ extern TPWRAPPER_API SecAccessFlags getDZFileScopePermissions(IEspContext& conte
+    return permission;
+}
+
+//Validate dropzone host and file path. Also validate dropzone file permission when the file path does not contain a file name.
```
The problem is inside the validateDropZoneHostAndPath() which is called by validateDropZoneReq() and not check isDropZoneRestrictionEnabled(). It should be fixed because the validateDropZoneHostAndPath() is also called by the FileSpray.FileList. I am working on it now.

wangkx commented:
@jakesmith please review my changes (submitted 3 hr ago. The checks are still not completed. Hopefully, they are successful.)

jakesmith commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -3023,23 +3023,18 @@ bool CFileSprayEx::onFileList(IEspContext &context, IEspFileListRequest &req, IE
-                    throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Access DropZone %s %s not allowed for user %s (permission:%s). Read Access Required.",
-                        dropZoneName, sPath.str(), context.queryUserId(), getSecAccessFlagName(permission));
+                if (isEmptyString(dropZoneName))
+                    dropZoneName= dropZone->queryProp("@name");
```
trivial/formattnig: spacing around '='

wangkx replied:
Fixed

jakesmith commented on esp/smc/SMCLib/TpCommon.cpp:
```diff
@@ -301,7 +306,34 @@ extern TPWRAPPER_API void validateDZFileScopePermissions(IEspContext& context, c
+
+    Owned<IPropertyTree> dropZone = getDropZoneAndValidateHostAndPath(dropZoneName, host, filePath);
+    if (dropZone)
+        dropZoneName= dropZone->queryProp("@name");
```
trivial/formatting: spacing around '='

wangkx replied:
Fixed

jakesmith reviewed (approved):
@wangkx - looks clearer, and ready. 2 trivial spacing issues.
Please address those and squash

wangkx commented:
@ghalliday please merge this PR.
